### ISSUE TITLE
plugin output error ES 5.4

### ISSUE THREAD

**Issue by 2803media** [logstash-plugins/logstash-input-elasticsearch#66]
Just upgrade to 5.4 and notice that json, csv... output in logstash make errors now and perfectly work on ES 5.3

exemple of conf file

```
input {
  elasticsearch {
	hosts => "http://localhost:9200"
	index => "sirene"
	query => '{"query": {"query_string" : {"query": "((DEPET:54 AND provider:sp_mairie) OR (DEPET:55 AND provider:sp_mairie) OR (DEPET:60 AND provider:sp_mairie))"}}}'
  }
}
output {
  csv {
	fields => ["nom", "street", "cp","nomcommune","nom_maire","siteweb","email"]
	path => "/home/data-prospection/public_html/jsondata/57cae745456ab85d4ff76a83ef3f1f0e/dt_0d83a4d79454b856ab249ca6496b17b1.csv"
  }
}
```

Errors

```
19:57:49.856 [[main]<elasticsearch] ERROR logstash.pipeline - A plugin had an unrecoverable error. Will restart this plugin.
  Plugin: <LogStash::Inputs::Elasticsearch hosts=>["http://localhost:9200"], index=>"sirene", query=>"{\"query\": {\"query_string\" : {\"query\": \"((DEPET:54 AND provider:sp_mairie) OR (DEPET:55 AND provider:sp_mairie) OR (DEPET:60 AND provider:sp_mairie))\"}}}", id=>"ed057865a1d953e8c84455e0e115b774f2045393-1", enable_metric=>true, codec=><LogStash::Codecs::JSON id=>"json_2b455f6f-f285-4213-bfe7-3ca311b027b0", enable_metric=>true, charset=>"UTF-8">, size=>1000, scroll=>"1m", docinfo=>false, docinfo_target=>"@metadata", docinfo_fields=>["_index", "_type", "_id"], ssl=>false>
  Error: [400] {"error":{"root_cause":[{"type":"illegal_argument_exception","reason":"Failed to parse request body"}],"type":"illegal_argument_exception","reason":"Failed to parse request body","caused_by":{"type":"json_parse_exception","reason":"Unrecognized token 'DnF1ZXJ5VGhlbkZldGNoBQAAAAAAAAC6Fm84Mm84SFpzU3VHSzdWMHdWQ3N3NGcAAAAAAAAAvRZvODJvOEhac1N1R0s3VjB3VkNzdzRnAAAAAAAAALsWbzgybzhIWnNTdUdLN1Ywd1ZDc3c0ZwAAAAAAAAC8Fm84Mm84SFpzU3VHSzdWMHdWQ3N3NGcAAAAAAAAAvhZvODJvOEhac1N1R0s3VjB3VkNzdzRn': was expecting ('true', 'false' or 'null')\n at [Source: org.elasticsearch.transport.netty4.ByteBufStreamInput@5039e667; line: 1, column: 457]"}},"status":400}
```
I run logstash 5.4.0
and I did this after the upgrade:

bin/logstash-plugin install logstash-input-elasticsearch

**Comment by jordansissel** [comment]
Confirmed. Logstash 5.3.0 and 5.4.0 both tested. Talking to ES 5.4.0:

```
% bin/logstash -e 'input { elasticsearch { index => foo } } output { null { } }'
[2017-05-10T12:10:11,381][ERROR][logstash.pipeline        ] A plugin had an unrecoverable error. Will restart this plugin.
  Plugin: <LogStash::Inputs::Elasticsearch index=>"foo", id=>"de68af58c20eb507df2323e1e31b6f1cec6984cd-1", enable_metric=>true, codec=><LogStash::Codecs::JSON id=>"json_b7ef4d87-d686-42f2-a543-98fc5408dba2", enable_metric=>true, charset=>"UTF-8">, query=>"{ \"sort\": [ \"_doc\" ] }", size=>1000, scroll=>"1m", docinfo=>false, docinfo_target=>"@metadata", docinfo_fields=>["_index", "_type", "_id"], ssl=>false>
  Error: [400] {"error":{"root_cause":[{"type":"illegal_argument_exception","reason":"Failed to parse request body"}],"type":"illegal_argument_exception","reason":"Failed to parse request body","caused_by":{"type":"json_parse_exception","reason":"Unrecognized token 'DnF1ZXJ5VGhlbkZldGNoBQAAAAAAAAOCFm9JTHlaRWpJVHgtejBYR3NZZ2RUMXcAAAAAAAADgBZvSUx5WkVqSVR4LXowWEdzWWdkVDF3AAAAAAAAA4MWb0lMeVpFaklUeC16MFhHc1lnZFQxdwAAAAAAAAOEFm9JTHlaRWpJVHgtejBYR3NZZ2RUMXcAAAAAAAADgRZvSUx5WkVqSVR4LXowWEdzWWdkVDF3': was expecting ('true', 'false' or 'null')\n at [Source: org.elasticsearch.transport.netty4.ByteBufStreamInput@5285fafa; line: 1, column: 457]"}},"status":400}
```

I can also reproduce with Logstash 5.3.0 Elasticsearch 5.3.0.

**Comment by jordansissel** [comment]
The http request looks like this:

```
GET /_search/scroll?scroll=1m HTTP/1.1.
Content-Type: application/json.
User-Agent: Faraday v0.9.2.
Accept: */*.
Connection: close.
Host: localhost:9200.
Content-Length: 228.

DnF1ZXJ5VGhlbkZldGNoBQAAAAAAAAEtFmx6c2hSRVRZVHh5ZGJCXzhZTTJISFEAAAAAAAABLxZsenNoUkVUWVR4eWRiQl84WU0ySEhRAAAAAAAAAS4WbHpzaFJFVFlUeHlkYkJfOFlNMkhIUQAAAAAAAAEwFmx6c2hSRVRZVHh5ZGJCXzhZTTJISFEAAAAAAAABMRZsenNoUkVUWVR4eWRiQl84WU0ySEhR
```

**Comment by jordansissel** [comment]
The problem is the content type is incorrect for the body of the request.

**Comment by jordansissel** [comment]
My theory is this:

* Elasticsearch 5.3.0 being more strict about the Content-Type header on requests
* The Ruby Elasticsearch client now defaulting to 'Content-Type: application/json'
* And how we are calling the scroll api

--- 

Reproducing in just ruby:

```
>> c = Elasticsearch::Client.new
>> c.search(index: "foo", body: nil, scroll: "1m", size: 10)
=> {"_scroll_id"=>"DnF1ZXJ5VGhlbkZldGNoBQAAAAAAAAQCFmx6c2hSRVRZVHh5ZGJCXzhZTTJISFEAAAAAAAAEAxZsenNoUkVUWVR4eWRiQl84WU0ySEhRAAAAAAAABAUWbHpzaFJFVFlUeHlkYkJfOFlNMkhIUQAAAAAAAAQEFmx6c2hSRVRZVHh5ZGJCXzhZTTJISFEAAAAAAAAEBhZsenNoUkVUWVR4eWRiQl84WU0ySEhR", "took"=>5, "timed_out"=>false, "_shards"=>{"total"=>5, "successful"=>5, "failed"=>0}, "hits"=>{"total"=>1, "max_score"=>1.0, "hits"=>[{"_index"=>"foo", "_type"=>"bar", "_id"=>"AVvzxTIPhuPq5JMzOkMf", "_score"=>1.0, "_source"=>{}}]}}

>> c.scroll(body: "DnF1ZXJ5VGhlbkZldGNoBQAAAAAAAAQCFmx6c2hSRVRZVHh5ZGJCXzhZTTJISFEAAAAAAAAEAxZsenNoUkVUWVR4eWRiQl84WU0ySEhRAAAAAAAABAUWbHpzaFJFVFlUeHlkYkJfOFlNMkhIUQAAAAAAAAQEFmx6c2hSRVRZVHh5ZGJCXzhZTTJISFEAAAAAAAAEBhZsenNoUkVUWVR4eWRiQl84WU0ySEhR", scroll: "1m")
Elasticsearch::Transport::Transport::Errors::BadRequest: [400] {"error":{"root_cause":[{"type":"illegal_argument_exception","reason":"Failed to parse request body"}],"type":"illegal_argument_exception","reason":"Failed to parse request body","caused_by":{"type":"json_parse_exception","reason":"Unrecognized token 'DnF1ZXJ5VGhlbkZldGNoBQAAAAAAAAQCFmx6c2hSRVRZVHh5ZGJCXzhZTTJISFEAAAAAAAAEAxZsenNoUkVUWVR4eWRiQl84WU0ySEhRAAAAAAAABAUWbHpzaFJFVFlUeHlkYkJfOFlNMkhIUQAAAAAAAAQEFmx6c2hSRVRZVHh5ZGJCXzhZTTJISFEAAAAAAAAEBhZsenNoUkVUWVR4eWRiQl84WU0ySEhR': was expecting ('true', 'false' or 'null')\n at [Source: org.elasticsearch.transport.netty4.ByteBufStreamInput@c94e6bf; line: 1, column: 457]"}},"status":400}
...
```

**Comment by jordansissel** [comment]
Possible bugs here:

* The Ruby Elasticsearch `scroll` method should check if the body is a string, and if so, set `content-type` appropriately (text/plain?)
* Maybe we've always been calling this incorrectly, and we shoudl call `.scroll(:body => {:scroll_id => scroll_id, ... })`

**Comment by jordansissel** [comment]
This bug impacts Logstash 5.3.0 and 5.4.0.  I have not tested other versions.

**Comment by jordansissel** [comment]
Plugin version 4.0.3 includes this fix. It was released on April 7 last month, but it was not included (by accident) in Logstash 5.4.0.

It will be included in Logstash 5.4.1.

To fix this for you, today, run the following command:

```
% bin/logstash-plugin update logstash-input-elasticsearch
Updating logstash-input-elasticsearch
Updated logstash-input-elasticsearch 4.0.2 to 4.0.3
```

**Comment by 2803media** [comment]
Thanks @jordansissel 

It's working now!
